server/dataset/cxg_dataset.py

```diff
@@ -271,10 +272,15 @@ def get_X_array(self, obs_mask=None, var_mask=None):
 
             nrows, obsindices = self.__remap_indices(X.shape[0], obs_mask, data.get("coords", data)["obs"])
             ncols, varindices = self.__remap_indices(X.shape[1], var_mask, data.get("coords", data)["var"])
-            densedata = np.zeros((nrows, ncols), dtype=self.get_X_array_dtype())
-            densedata[obsindices, varindices] = data[""]
-
-            return densedata
+            if obsindices.size / nrows > 0.5:
```

Comment:
I think this is the wrong place to put this logic.  The decision being made is how to encode in a flatbuffer, which belongs inside the flatbuffer encoder (eg, in matrix.py), not in this code which is used for many purposes (eg, internal computation like diffex, etc).

Put another way, the matrix encoder should introspect the data and decide on the best FB encoding to use.

Reply:
One concern I have is that returning a dense array which then needs to be converted into a sparse array would add >100 milliseconds to the request for large (>4m) datasets. How about we add an argument to `get_X_array` called `return_dense` which would return dense data regardless of density for functions that need it, like diffex.

Reply:
Or we could do the opposite and add an `allow_sparse` argument and `density=0.5` argument to determine if `get_X_array` is allowed to return a sparse data type based on the specified density cutoff. That way, default behavior is how `get_X_array` behaved previously.

Reply:
Perhaps the better trade is time of network transfer vs time to encode/decode.  It needs to be a net win across both.

One concrete suggestion:  only use the complicated encoding if:
* the array is large
* and it has low density
If both are true, then the smaller size is well worth the CPU time.  If the array is already small, the complicated encoding may not be worth it.

You will need to do some experiments to find the right threshold.